### ISSUE TITLE
[SUGGESTION] can we add a way to use attributes?

### ISSUE THREAD

**Issue by farmerpiki** [hsutter/cppfront#1252]
I am not aware of any way to hint the branch predictor in the new syntax

I'd like to be able to use attributes, example cpp1 code:

```
if(condition) [[likely]] {
 ...
} else {
 ...
}
```

**Comment by hsutter** [comment]
Thanks for the suggestion! I haven't implemented attributes yet, but intend to.

**Comment by zaucy** [comment]
There might be some overlap with meta-functions and attributes. Maybe C++ attributes could be implemented as meta functions if they are allowed on if statements, expressions etc.

```cpp2
if condition @likely {
  // ...
} else {
  // ...
}

// or maybe it should be before the if
@likely if condition {
  // ...
} else {
  // ...
}
```

**Comment by SebastianTroy** [comment]
What do you envision the meta function looking like in the case of @likely?


On 21 August 2024 19:30:42 Ezekiel Warren ***@***.***> wrote:

There might be some overlap with meta-functions and attributes. Maybe C++ attributes could be implemented as meta functions if they are allowed on if statements, expressions etc.

if condition @likely {
  // ...
} else {
  // ...
}

// or maybe it should be before the if
@likely if condition {
  // ...
} else {
  // ...
}

**Comment by zaucy** [comment]
perhaps `compiler_services` could expose an "add/prepend cpp1" so the likely metafunction could look something like this:

```
likely: (s: cpp2::meta::statement) = {
  if s.is_if_statement() {
    s.append_cpp1("[[likely]]");
  } else {
    // TODO: add likely for other branches
    error("cannot add likely to non-if statement");
  }
}
```

**Comment by farmerpiki** [comment]
> perhaps `compiler_services` could expose an "add/prepend cpp1" so the likely metafunction could look something like this:
> 
> ```
> likely: (s: cpp2::meta::statement) = {
>   if s.is_if_statement() {
>     s.append_cpp1("[[likely]]");
>   } else {
>     // TODO: add likely for other branches
>     error("cannot add likely to non-if statement");
uhm yes you can, you can add `[[likely]]` to switch statements(this just tells the compiler to test for those marked with likely first, it would probably work the same if you put them first in your code, but it might make it harder for a human to reason about, for example if 4 and 7 are the most likely cases for an int... and you put them first it might seem to people reading the code that some cases were skipped). In many cases you should add `[[unlikely]]` to the `default` switch case
>   }
> }
> ```

**Comment by farmerpiki** [comment]
we actually should have both `likely` and `unlikely` ... likely means prioritize this case and unlikely means put this code in another bucket because it might never get called and it allows for keeping the happy flow closer together in cache so that when all goes well it goes faster

**Comment by farmerpiki** [comment]
this suggestion was meant to be more generic though... I want to be able to add attributes to functions as well... and a way to add them so that it just works no matter the compiler or attribute I want to be able to use

`[[gnu::target_clones("default", "sse4.2", "avx", "avx2")]]`

and others as well

**Comment by dutkalex** [comment]
Using metafunctions to add attributes makes a lot of sense IMO:
- it fits nicely in the "reducing the concept count" and "simplification through generalization" principles
- attributes can become a library feature rather than a core language thing, which enables the use of non-standard (user-defined or compiler-dependant) ones for special purposes

**Comment by DyXel** [comment]
When all you have is a hammer, everything looks like a nail. I personally don't believe metafunctions / metaprograms are the right approach for attributes, as they are supposed to be hints for the implementation, rather than having big impact on the program you are writing (at least that's my perception).

**Comment by dutkalex** [comment]
I agree that it makes sense to distinguish things that are just optimization hints which can be ignored from metafunctions which offer guarantees.
However, coming from an HPC background, I have observed that:
- people tend to rely on certain optimizations actually being implemented (I know that this is not really a good argument to guide language design but non-experts tend to think that if "it's written then it's guaranteed" and I think it is worth taking that into account)
- in order to leverage compiler-specific attributes and attribute-like language extensions (``__device__`` annotations for gpu programming for example) in a portable way, the answer is almost always to use macros because there is no other way to insert logic before stamping a function with an attribute. Cpp2 offers a natural alternative with metafunctions, which can encapsulate the raw attributes into solid domain specific abstractions. I believe that in this context most code will never use raw attributes and instead rely on more expressive metafunctions, hence the idea that maybe attributes should simply be primitive metafunctions.

**Comment by dutkalex** [comment]
The line between what is just an optimization hint and should have first class support can be blurry. For example, one can argue that loop unrolling is just an optimization and that this should be left to the compiler (maybe with the ability to add an attribute as a hint).
But this currently leads to loads of TMP hacks to achieve the same purpose with guarantees. I personally have in my codebase something along these lines for this exact purpose, because it extends the expressive power of the language:
```cpp
static_for< 0, N >( [&]( auto i ){
  some_regular_function_call( i.value );
  some_template_function_call< i.value >();
} );
```

**Comment by SebastianTroy** [comment]
I don't see how a library can implement some of these features, at some point there must be some indication to the compiler, an [[attribute]] so to speak, is burying that in a meta function better than declaring it upfront?


On 22 August 2024 13:25:07 Alex Dutka ***@***.***> wrote:

Using metafunctions to add attributes makes a lot of sense IMO:

  *   it fits nicely in the "reducing the concept count" and "simplification through generalization" principles
  *   attributes can become a library feature rather than a core language thing, which enables the use of non-standard (user-defined or compiler-dependant) ones for special purposes

**Comment by dutkalex** [comment]
> I don't see how a library can implement some of these features

I don't see why we must have attributes in cpp2 to emit cpp1 code with attributes. Am I missing something there?

**Comment by SebastianTroy** [comment]
Conceptually, at some point, there must be something that the transpiler, reads and understands as an attribute, rather than code. I don't believe that reserving various meta function names to replace attributes is a good course of action.


On 22 August 2024 15:46:15 Alex Dutka ***@***.***> wrote:

I don't see how a library can implement some of these features

I don't see why we must have attributes in cpp2 to emit cpp1 code with attributes. Am I missing something there?

**Comment by zaucy** [comment]
not to detract from C++ attributes too much here, but to expand the idea of "emitting cpp1" in a metafunction (which would cover C++ attributes) - the unreal engine has a sort of "attribute" that they decorate their classes and functions with that I think having an "emitting cpp1" feature could cover.

for instance you could write an unreal engine class in cpp2 like this:

```cpp2
AMyActor: @uclass type = {
  this: AActor;
  // ...
};
```

that would lower to this

```cpp
UCLASS()
class AMyActor : AActor {
  GENERATED_BODY()
};
```

**Comment by DyXel** [comment]
Ideally, the metaclass would expand to whatever the macro expands to, not to the macro itself. Same idea with `Q_OBJECT` in Qt. Herb touched on this in one of his talks, I don't remember exactly which one though 😅

**Comment by zaucy** [comment]
> Ideally, the metaclass would expand to whatever the macro expands to, not to the macro itself

Totally agree, but it would be a really nice stepping stone. Especially for the unreal example above since if it could emit what I described it could be used today since it could use existing tools (unreal header tool, unreal build tool, etc.)

**Comment by JohelEGP** [comment]
I've been waiting on `[[no_unique_address]]`.
> I also use this to have `[[no_unique_address]]` members:
> ```C++
> #define QTY_NAME_MEM_TYPE [[no_unique_address]] name_type
> ```
>
> -- <https://github.com/hsutter/cppfront/discussions/658#discussioncomment-6899974>

**Comment by JohelEGP** [comment]
I agree that we shouldn't abuse metafunctions.

The likes of `@java_interface` and `@qt::moc` are not only fine,
they are one of the intended audience of metafunctions.
Even if they don't modify the type, but only read it to write some output.

There are other features we don't have in Cpp2 yet.
These aren't Cpp1 attributes, but could be expressed in their Cpp2 syntax or as metafunctions.
- `@dll_visible` from <https://github.com/hsutter/cppfront/pull/907#issuecomment-1871495998>.
- Cpp1's `= delete("with a reason");`.
- Trivial functions (#547).
